File: pa1/List.java
/*
Zachary Dutra
zdutra
1581789
pa1
*/

class Node{
    int data;
    Node next;

    Node(int data){
        this.data = data;
        this.next = null;
    }

    public String toString(){
        return String.valueOf(data);
    }
}

class List{
    Node front;
    Node back;
    Node cursor;
    int size;
    int index;
    List(){ // Creates a new empty list.
         this.front = this.back = this.cursor = null;
         this.size  = 0;
         this.index = -1;
    }       
    // Access functions
    int length(){       // Returns the number of elements in this List.
            return this.size;
        }
    int index(){            // If cursor is defined, returns the index of the cursor element,
            if(this.cursor==null){          // otherwise returns -1.
                    return -1;
            }
            else{
                return this.index;
            }           
        }
    int front(){
            return front.data;
        }           // Returns front element. Pre: length()>0
    int back(){
            return back.data;
        }           // Returns back element. Pre: length()>0
    int get(){
            if(this.index == -1||size == 0){
                return -1;
            }
            else{
                return this.cursor.data;
            }
        }           // Returns cursor element. Pre: length()>0, index()>=0
    boolean equals(List L){
            Node temp = this.front;
            Node temp2 = L.front;
            if(this.size!=L.size){
                    return false;
            }
            for(int i = 0; i < size; i++){
                if(temp.data!=temp2.data){
                    return false;
                }
                temp = temp.next;
                temp2 = temp2.next;
            }
            return true;
        }           // Returns true if and only if this List and L are the same
                    // integer sequence. The states of the cursors in the two Lists
                    // are not used in determining equality.
                    // Manipulation procedures
    void clear(){
            this.front = this.back = this.cursor = null;
            this.size = 0;
            this.index = -1;
        }            // Resets this List to its original empty state.
    void moveFront(){
            if(size==0){
                return;
            }
            this.cursor = this.front;
            this.index = 0;
            return;
        }           // If List is non-empty, places the cursor under the front element,
                    // otherwise does nothing.
    void moveBack(){
            if(size==0){
                return;
            }
            this.cursor = this.back;
            this.index = size-1;
            return;
    }               // If List is non-empty, places the cursor under the back element,
                    // otherwise does nothing.
    void movePrev(){
            if(this.cursor==null){
                return;
            }
            else if(this.cursor==this.front){
                this.cursor = null;
                this.index = -1;
                return;
            }
            else{
                Node temp = this.front;
                while(temp.next!=this.cursor){
                    temp = temp.next;
                }
                this.cursor = temp;
                this.index--;
                return;
            }
    }               // If cursor is defined and not at front, moves cursor one step toward
                    // front of this List, if cursor is defined and at front, cursor becomes
                    // undefined, if cursor is undefined does nothing.
    void moveNext(){
            if(this.cursor == this.back){
                this.cursor = null;
                this.index = -1;
                return;
            }
            else if(this.cursor == null){
                return;
            }
            else{
                this.cursor = this.cursor.next;
                this.index++;
                return;
            }
    }               // If cursor is defined and not at back, moves cursor one step toward
                    // back of this List, if cursor is defined and at back, cursor becomes
                    // undefined, if cursor is undefined does nothing.
    void prepend(int data){
            if(size == 0){
                Node temp = new Node(data);
                this.front = this.back = temp;
                this.size++;
                return;
            }
            else{
                Node temp = new Node(data);
                temp.next = this.front;
                this.front = temp;
                this.size++;
                return;
            }
    }                // Insert new element into this List. If List is non-empty,
                     // insertion takes place before front element.
    void append(int data){
            if(size==0){
                Node temp = new Node(data);
                this.front = this.back = temp;
                this.size++;
                return;
            }
            else{
                Node temp = new Node(data);
                this.back.next = temp;
                this.back = temp;
                this.size++;
                return;
            }

    }                // Insert new element into this List. If List is non-empty,
                    // insertion takes place after back element.
    void insertBefore(int data){
        if(this.index < 0 || this.size == 0){
            return;
        }
        else if(this.index == 0){
                prepend(data);
                this.size++;
                return;
        }
        else{
            Node temp = new Node(data);
            Node temp2 = this.front;
            while(temp2.next!=this.cursor){
                temp2 = temp2.next;
            }
            temp2.next = temp;
            temp.next = this.cursor;
            this.size++;
            return;
        }
    }               // Insert new element before cursor.
                    // Pre: length()>0, index()>=0
    void insertAfter(int data){
        if(this.index < 0 || this.size == 0){
            return;
        }
        else if(this.index == size-1){
            append(data);
            this.size++;
            return;
        }
        else{
            Node temp = new Node(data);
            temp.next = this.cursor.next;
            this.cursor.next = temp;
            this.size++;
            return;
        }
    }               // Inserts new element after cursor.
                    // Pre: length()>0, index()>=0
    void deleteFront(){
        if(size == 0){
            return;
        }
        this.front = this.front.next;
        size--;
        return;
    }               // Deletes the front element. Pre: length()>0
    void deleteBack(){
        if(size==0){
            return;
        }
        Node deltemp = this.front;
        while(deltemp.next!=this.back){
            deltemp = deltemp.next;
        }
        deltemp.next = null;
        this.back = deltemp;
        size--;
        return;
    }               // Deletes the back element. Pre: length()>0

    void delete(){
        if(this.index < 0 || size == 0){
            return;
        }
        else{
            Node temp = this.front;
            while(temp.next!=this.cursor){
                temp = temp.next;
            }
            temp.next = this.cursor.next;
            this.cursor = this.cursor.next = null;
            this.index = -1;
            size--;
            return;
        }
    }               // Deletes cursor element, making cursor undefined.
                    // Pre: length()>0, index()>=0
                    // Other methods
    public String toString(){
        Node temp = this.front;
        String[] output_A = new String[this.size];
        String output_S = "";
        for(int i = 0; i < this.size; i++){
            output_A[i] = String.valueOf(temp.data);
            temp = temp.next;
        }
        for(int j = 0; j < output_A.length; j++){
            output_S = output_S + output_A[j] + " ";
        }
        return output_S;
    }               // Overrides Object's toString method. Returns a String
                    // representation of this List consisting of a space
                    // separated sequence of integers, with front on left.
    List copy(){
        List newList = new List();
        newList.front = this.front;
        Node temp = this.front;
        Node newTemp = newList.front;
        newList.size = this.size;
        while(temp != null){
            temp = temp.next;
            newTemp.next = temp;
            newTemp = newTemp.next;
        }
        return newList;
    }               // Returns a new List representing the same integer sequence as this
                    // List. The cursor in the new list is undefined, regardless of the
                    // state of the cursor in this List. This List is unchanged.
}

File: pa4/List.c
//-----------------------------------------------------------------------------
// List.c
// Zachary Dutra
// zdutra
// 1581789
// pa2
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// structs --------------------------------------------------------------------

// private NodeObj type
typedef struct NodeObj{
   int data;
   struct NodeObj* next;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private QueueObj type
typedef struct ListObj{
   Node front;
   Node back;
   Node cursor;
   int length;
   int index;
} ListObj;


// Constructors-Destructors ---------------------------------------------------

// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

// newList()
// Returns reference to new empty List object.
List newList(void){
   List L;
   L = malloc(sizeof(ListObj));
   L->front = L->back = L->cursor = NULL;
   L->length = 0;
   L->index = -1;
   return(L);
}


// freeList()
// Frees all heap memory associated with List *pL, and sets *pL to NULL.S
void freeList(List* pL){
   if(pL!=NULL && *pL!=NULL) { 
      while( !isEmpty(*pL) ) { 
         deleteFront(*pL); 
      }
      free(*pL);
      *pL = NULL;
   }
}


// Access functions -----------------------------------------------------------

// getFront()
// Returns the value at the front of L.
// Pre: !isEmpty(L)
int front(List L){
   if( L == NULL ){
      printf("List Error: calling front() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling front() on empty List reference\n");
      exit(1);
   }
   return(L->front->data);
}

//back()
//Returns value at the back of the List
int back(List L){
   if( L == NULL ){
      printf("List Error: calling back() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling back() on empty List reference\n");
      exit(1);
   }
   return(L->back->data);
}

// Length()
// Returns the length of L.
int length(List L){
   if( L==NULL ){
      printf("List Error: calling length() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      return 0;
   }
   return(L->length);
}

//index()
//Returns current List index
int index (List L){
   if( L==NULL ){
      printf("List Error: calling index() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      return -1;
   }
      return (L->index);
}

// get()
// returns data at current cursor node
int get(List L){
   if( L==NULL ){
      printf("List Error: calling get() on NULL List reference\n");
      exit(1);
   }
   else if( L->cursor == NULL){
      return -1;
   }
   return (L->cursor->data);
}
// equals()
// returns true (1) if A is identical to B, false (0) otherwise
int equals(List A, List B){
   int eq = 0;
   Node N = NULL;
   Node M = NULL;

   if( A==NULL || B==NULL ){
      printf("List Error: calling equals() on NULL List reference\n");
      exit(1);
   }

   eq = ( A->length == B->length );
   N = A->front;
   M = B->front;
   while( eq && N!=NULL){
      eq = (N->data==M->data);
      N = N->next;
      M = M->next;
   }
   return eq;
}

// isEmpty()
// Returns true (1) if Q is empty, otherwise returns false (0)
int isEmpty(List L){
   if( L==NULL ){
      printf("List Error: calling isEmpty() on NULL List reference\n");
      exit(1);
   }
   return(L->length==0);
}


// Manipulation procedures ----------------------------------------------------

//clear()
//resets input list on initial values
void clear(List L){
   L->front = L->back = L->cursor= NULL;
   L->index = -1;
   L->length = 0;
   return;
}

// append()
// Places new data element at the end of L
void append(List L, int data)
{
   Node N = newNode(data);

   if( L==NULL ){
      printf("List Error: calling Enqueue() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ) { 
      L->front = L->back = N; 
   }else{ 
      L->back->next = N; 
      L->back = N; 
   }
   L->length++;
   return;
}

// prepend()
// Places new data element at the beginning of L
void prepend(List L, int data)
{
   Node N = newNode(data);

   if( L==NULL ){
      printf("List Error: calling Enqueue() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ) { 
      L->front = L->back = N; 
   }else{ 
      N->next = L->front;
      L->front = N; 
   }
   L->length++;
   if(L->index != -1){
      L->index++;
   }
   return;
}
//moveFront()
// moves cursor to the front
void moveFront(List L){
   if(L->length==0){
      return;
   }
   else{
      L->cursor = L->front;
      L->index = 0;
      return;
   }
}

//moveBack()
//moves cursor to the back
void moveBack(List L){
   if(L->length==0){
      return;
   }
   else{
      L->cursor = L->back;
      L->index = L->length - 1;
      return;
   }
}

//movePrev()
//moves cursor to previous element
void movePrev(List L){
   if(L->cursor == L->front){
      L->cursor = NULL;
      L->index = -1;
      return;
   }
   else{
      Node N = L->front;
      while(N->next != L->cursor){
         N = N->next;
      }
      L->cursor = N;
      L->index--;
      return;
   }
}

//moveNext()
//moves cursor to next element
void moveNext(List L){
   if(L->cursor == L->back){
      L->cursor = NULL;
      L->index = -1;
      return;
   }
   else{
      L->cursor = L->cursor->next;
      L->index++;
      return;
   }
}

//insertBefore()
//inserts new element before the cursor
void insertBefore(List L, int data){
   if(L->length == 0 || L->cursor == L->front){
      prepend(L, data);
      return;
   }
   else if (L->cursor == NULL) {
      return;
   }
   else{
      Node N = newNode(data);
      Node temp = L->front;
      while(temp->next != L->cursor){
         temp = temp->next;
      }
      temp->next = N;
      N->next = L->cursor;
      L->length++;
      L->index++;
      return;
   }
}

//insertAfter()
//insert new element after the cursor
void insertAfter(List L, int data){
   if(L->length == 0){
      prepend(L, data);
      return;
   }
   else if(L->cursor == L->back){
      append(L, data);
      return;
   }
   else if (L->cursor == NULL) {
      return;
   }
   else{
      Node N = newNode(data);
      N->next = L->cursor->next;
      L->cursor->next = N;
      L->length++;
      return;
   }
}
// deleteFront()
// Deletes element at front of L
// Pre: !isEmpty(L)
void deleteFront(List L){
   Node N = NULL;
   if( L==NULL ){
      printf("List Error: calling deleteFront() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling deleteFront() on an empty List\n");
      exit(1);
   }
   if(L->cursor == L->front){
      L->cursor = NULL;
      L->index = -1;
   }
   N = L->front;
   if( length(L)>1 ) { 
      L->front = L->front->next; 
   }
   else{ 
      L->front = L->back = NULL; 
   }
   L->length--;
   if(L->index != -1){
      L->index--;
   }
   freeNode(&N);
   return;
}

// deleteBack()
// Deletes element at back of L
// Pre: !isEmpty(L)
void deleteBack(List L){
   Node N = L->front;

   if( L==NULL ){
      printf("List Error: calling deleteBack() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling deleteBack() on an empty List\n");
      exit(1);
   }
   if(L->cursor == L->back){
      L->cursor = NULL;
      L->index = -1;
   }
   if( length(L)>1 ) { 
      while(N->next != L->back){
         N = N->next;
      }
      N->next = NULL;
      L->back = N;
      L->length--;
      freeNode(&N);
      return;
   }
   else{ 
      L->front = L->back = NULL; 
       L->length--;
      freeNode(&N);
      return;
   }
}

//delete()
//deletes current cursor node
void delete(List L){
      if(L->cursor==NULL){
         return;
      }
      if(L->cursor == L->front){
         deleteFront(L);
         return;
      }
      if(L->cursor == L->back){
         deleteBack(L);
         return;
      }
      else{
         Node N = L->front;
         while(N->next != L->cursor){
            N = N-> next;
         }
         N->next = L->cursor->next;
         L->cursor = NULL;
         L->index = -1;
         L->length--;
         return;
      }
}


// Other Functions ------------------------------------------------------------

// printList()
// Prints data elements in L on a single line to out.
void printList(FILE* out, List L){
   Node N = NULL;

   if( L==NULL ){
      printf("List Error: calling printList() on NULL List reference\n");
      exit(1);
   }

   for(N = L->front; N != NULL; N = N->next){
      fprintf(out, "%d ", N->data);
   }
   fprintf(out, "\n");
}

//copyList()
//returns a new list and copies input list to new list
List copyList(List L){
   List new_list = newList();
   new_list->front = L->front;
   Node temp = L->front;
   Node newTemp = new_list->front;
   new_list->length = L->length;
   while(temp != NULL){
      temp = temp->next;
      newTemp->next = temp;
      newTemp = newTemp->next;
   }
   return new_list;
}
